server/apps/archive/archive.py

```diff
@@ -412,6 +413,78 @@ def can_edit(self, item, user_id):
+        """
+
+        assert doc[config.CONTENT_STATE] == 'spiked', \
+            "Article state is %s. Only Spiked Articles can be removed" % doc[config.CONTENT_STATE]
+
+        doc_id = str(doc[config.ID_FIELD])
+        super().delete_action({config.ID_FIELD: doc_id})
+        get_resource_service('archive_versions').delete(lookup={versioned_id_field(): doc_id})
+
+        packages = self.get_packages(doc_id)
+        if packages.count() > 0:
+            processed_packages = []
+            for package in packages:
+                if str(package[config.ID_FIELD]) not in processed_packages:
+                    processed_packages.extend(self.remove_refs_in_package(package, doc_id, processed_packages))
+
+    def get_packages(self, doc_id):
+        """
+        Retrieves if an article identified by doc_id is referenced in a package.
+        :return: articles of type composite
+        """
+
+        query = {'query': {'filtered': {'filter': {'and': [{'term': {'type': 'composite'}}]},
+                                        'query': {
+                                            'match': {'groups.refs.guid': {'query': doc_id, 'operator': 'AND'}}}
+                                        }}}
+
+        request = ParsedRequest()
+        request.args = {'source': json.dumps(query)}
+
+        return get_resource_service(SOURCE).get(req=request, lookup=None)
+
+    def remove_refs_in_package(self, package, ref_id_to_remove, processed_packages=None):
+        """
+        Removes residRef referenced by ref_id_to_remove from the package associations and returns the package id.
+        Before removing checks if the package has been processed. If processed the package is skipped.
+        :return: package[config.ID_FIELD]
+        """
+
+        groups = package['groups']
+
+        if processed_packages is None:
+            processed_packages = []
+
+        sub_package_ids = [ref['guid'] for group in groups for ref in group['refs'] if ref.get('type') == 'composite']
+        for sub_package_id in sub_package_ids:
+            if sub_package_id not in processed_packages:
+                sub_package = self.find_one(req=None, _id=sub_package_id)
+                return self.remove_refs_in_package(sub_package, ref_id_to_remove)
+
+        new_groups = [{'id': group['id'], 'role': group['role'],
+                       'refs': [ref for ref in group['refs'] if ref.get('guid') != ref_id_to_remove]}
+                      for group in groups]
+        new_root_refs = [{'idRef': group['id']} for group in new_groups if group['id'] != 'root']
+
+        for group in new_groups:
+            if group['id'] == 'root':
+                group['refs'] = new_root_refs
+                break
+
+        updates = {config.LAST_UPDATED: utcnow(), 'groups': new_groups}
+        resolve_document_version(updates, SOURCE, 'PATCH', package)
+
+        self.patch(package[config.ID_FIELD], updates)
+        insert_into_versions(id_=package[config.ID_FIELD])
```

Comment:
Maybe we should move this into PackagesService for this to happen on all package changes.